### ISSUE TITLE
Property patterns and dynamic

### ISSUE THREAD

**Issue by svick** [dotnet/roslyn#28933]
**Version Used**: SharpLab for branch features/recursive-patterns from 24 Jul 2018 (commit b364cfc)

**Steps to Reproduce**:

Try to compile [the following code](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRIMaOQSwG4SoAOsMECAdsgD4ACATAAQDCAsAFADeHjvjtAFkYBZABQATAJ4UoAWxwZGAewCUPPt3Z9tjAPqMAvMsY5kjTowBiSpSEYCmAXwDc63o46OgA===):

```c#
class C
{
    void M(dynamic o)
    {
        _ = o is { Foo: 42 };
    }
}
```

**Expected Behavior**:

Since code like `o is { Foo: 42 }` normally compiles to the equivalent of `o != null && o.Foo == 42` and since that code is valid for `o` of type `dynamic`, I would expect the above method to compile.

**Actual Behavior**:

> error CS0117: 'dynamic' does not contain a definition for 'Foo'

I think the error message is somewhat confusing, because the whole point of `dynamic` is that it's not checked at compile time whether a member exists.

---

As usual with work-in-progress features, it's not clear to me if this is by design, not implemented yet, just a confusing error message or an actual bug, so feel free to handle this issue accordingly.

**Comment by gafter** [comment]
I agree the diagnostic could be improved. The behavior is intentional. Pattern-matching is a different, non-exception-throwing way to dynamically check the shape of data. Dynamic throws. These are incompatible.

**Comment by svick** [comment]
I understand that it wouldn't make much sense to use this syntax to check whether the given object has a property named `Foo`. But if I know that it has that property, I have to use `dynamic` to access it, and I only want to check whether the value of that property is `42`, then I think using property pattern would make sense here.

Though I understand that might not be worth doing, especially since people might expect that it's checking whether the object has the right shape.

**Comment by HaloFour** [comment]
IMO I'd rather recursive patterns not be permitted on `dynamic` at all for now so that it can leave the possibility of adding proper language support in the future when it might be possible to test the shape of dynamic instances.  I assume that would involve modifications to DLR?
